src/node_contractor.rs

```diff
@@ -133,7 +131,7 @@ mod tests {
         g.add_edge(0, 3, 1);
         g.add_edge(3, 2, 1);
         let shortcuts = calc_shortcuts(&mut g, 1);
-        assert_eq!(0, shortcuts.len());
+        assert_eq!(1, shortcuts.len());
```

Comment:
There should be no shortcut here: When we contract node 1 (or calculate the shortcuts as here) we should find 0->3->2 as witness path and we can simply remove node 1 without adding any shortcut. The reason you had to adjust this test is probably that you no longer set node 1 to be the 'avoid_node' during the witness search. This way we will immediately find the connection 0->1->2 and then add the shortcut 0->2.

Reply:
So for me the question here is: How do you avoid such unnecessary shortcuts if you do not disable the contracting node for the witness search?

Reply:
hmm true
using the `avoid_node` function will definitely create a shortcut in the test: `calc_shortcuts_witness_via_center`

Reply:
Not if we sort the edge-pairs ;)

Reply:
for the example:
in the future you would contract node 1 eventually
having two shortcuts with the same in an out nodes, but consisting of different edges. this would then be deduplicated, if I understand you correctly. So yes this is temporarily bad, but will resolve in the future.

Reply:
This I do not understand yet: What do you mean 'in the future'? We are contracting node 1 here and after the contraction the method you propose here yields this graph

```rust
// 0 -----> 2
//  \->3->/
```

Then it depends which node will be contracted next. If we contract 3 next it will be like you said the 0->2 edge will be deduplicated. But if we contract node 2 next for example edge 0->2 will exist in the final graph even though its not needed.

Reply:
the explanation for this is below with the heuristic